eos-split-flatpak-repo

```diff
@@ -219,6 +257,20 @@ def gather_refs(repo, os_remotes, flatpak_remotes):
     return os_refs, flatpak_refs, other_refs
 
 
+def record_repo_stats(repo, stats):
+    """Record statistics for repository data"""
+    logger.info('Recording repository %s stats', repo.path)
+
+    _, objects = repo.list_objects(
+        OSTree.RepoListObjectsFlags.ALL |
+        OSTree.RepoListObjectsFlags.NO_PARENTS
+    )
+    stats.num_objects = len(objects)
```

Comment:
Just to be up front about it, getting the number of objects here is a bit expensive. OSTree enumerates all the object files and then serializes each name into a GVariant. I'm not sure we really need this information, so I just want to be clear it's not free.

Reply:
I timed this on my Mission One (5400 rpm disk):

```
>>> x = time.monotonic(); print(len(repo.list_objects( OSTree.RepoListObjectsFlags.ALL | OSTree.RepoListObjectsFlags.NO_PARENTS )[1])); y = time.monotonic(); print(x, y, y-x)
327813
832.414858406 900.593506899 68.17864849299997
```

(This is after the split had completed but I think before the repo has been pruned.)

So that's 68 seconds to enumerate all the objects. I waited a few minutes (during which time the computer was otherwise unused but I could hear its disk buzzing continuously, looks like it was the updater) and re-ran the same invocation, 66 seconds.

Maybe there would be some cache-warming effect where the subsequent "hardlink all the things" operation will be slightly quicker, but still I fear this will actually make the problem we are trying to quantify and solve worse! I'd leave it out.

Reply:
OK, here's an idea: Run `cp` with `--verbose` and count the number of lines of output. Of course this includes directories too but on my system there are 514 directories and 928,349 non-directories across the ostree & flatpak repos so it's a reasonable approximation. Of course this would also come with a performance cost…

Reply:
Thanks for testing, that's too painful.

The simple way to do this with counting lines from `--verbose` would be to run with `stdout=PIPE` and then split it. However, that would buffer all of the output into memory, which could be a lot. Another way would be to write to a file then open and read lines in a more controlled way. Or use `stdout.PIPE` without `communicate`. Or `cp -v | wc -l` with `shell=True`, ensuring that shell was `bash` and you were checking the return value of `cp` correctly. I found https://stackoverflow.com/questions/845058/how-to-get-line-count-of-a-large-file-cheaply-in-python, which was fascinating. On my system none of the python approaches even got close to `wc`.

Reply:
Yet another option is `find repo/objects -type f` or some python equivalent. It's slightly cheaper than `ostree_repo_list_objects` because the names aren't being serialized into variants, but you still to enumerate all the paths. However, this is a little faster because it avoids `stat`ing all the objects since we don't care about that:

```
def listcount(repo):
    objsdir = os.path.join(repo, 'objects')
    nobjs = 0
    for entry in os.scandir(objsdir):
        if not entry.is_dir():
            continue
        nobjs += len(os.listdir(entry.path))
    return nobjs
```

This is basically using the knowledge of how `objects` is broken up into a two level content addressed storage. It was faster than `glob`, `walk` and `find | wc -l` on my system.

Reply:
Yeah, wow, `ostree_repo_list_objects` is like 500x slower than my fastest `listdir` approach. From https://phabricator.endlessm.com/P868:

```
$ ./countobjs.py /var/lib/flatpak/repo
globcount: 308969: 0.5490682479576208
walkcount: 308961: 0.2801525180111639
listdircount: 308969: 0.17220832197926939
listobjscount: 308815: 5.436755720991641
findcount: 300130: 0.46448501001577824
```

Would be useful if you could try that out on your spinning disk. I have a feeling it would still be faster than having `cp` write a ton of output to stdout and then read it all. Or maybe we just scrap all of this as too expensive...

Reply:
OK, I'll test this tonight/tomorrow but sadly I'm miles from my Mission One today and can't SSH to it… :)

Reply:
I tracked down my Mission spinning disk, so I'll try it out myself today.

Reply:
I tested this out and the `listdir` method took less than 1 second while the `list_objects` method took 54 seconds. I think I'll go with the `listdir` method...